plan/plans/join.go

```diff
@@ -187,20 +208,35 @@ func (r *JoinPlan) nextRightJoin(ctx context.Context) (row *plan.Row, err error)
 		if rightRow == nil || err != nil {
 			return nil, errors.Trace(err)
 		}
-		err = r.findMatchedRows(ctx, rightRow, true)
+
+		tempExpr := r.On.Clone()
```

Comment:
I tried to merge them in one function before, but that result code is much more complex and failed to pass tests.